zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,195 @@
+#![allow(non_snake_case)]
+
+use crate::errors::VMError;
+use crate::signature::multikey::Multikey;
+use crate::signature::musig::*;
+use crate::signature::VerificationKey;
+use crate::transcript::TranscriptProtocol;
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Nonce(Scalar);
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+#[derive(Copy, Clone, Debug)]
+pub struct Siglet(Scalar);
+
+pub struct PartyAwaitingPrecommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    R_i: NonceCommitment,
+}
+
+pub struct PartyAwaitingCommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    nonce_precommitments: Vec<NoncePrecommitment>,
+}
+
+pub struct PartyAwaitingSiglets {
+    multikey: Multikey,
+    c: Scalar,
+    nonce_commitments: Vec<NonceCommitment>,
+}
+
+impl NonceCommitment {
+    fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R_i", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+}
+
+impl PartyAwaitingPrecommitments {
+    pub fn new(
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        x_i: PrivKey,
+        multikey: Multikey,
+    ) -> (Self, NoncePrecommitment) {
+        let mut rng_transcript = transcript.clone();
+        rng_transcript.commit_scalar(b"x_i", &x_i.0);
+        let mut rng = rng_transcript.build_rng().finalize(&mut rand::thread_rng());
+
+        // Generate ephemeral keypair (r_i, R_i). r_i is a random nonce.
+        let r_i = Nonce(Scalar::random(&mut rng));
+        // R_i = generator * r_i
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i.0);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript: transcript.clone(),
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
```

Comment:
do we need to receive our own precommitment at a correct index here?

Reply:
we do not necessarily have to receive our own precommitment a a "correct" index, but we do have to receive it at the same index that we receive our commitment at, in the next step. 
(we could also assume that the dealer sends all of the precommitments and commitments _except_ for those already owned by the party, but it just requires more sorting on the dealer's part so I started with the most simple case first.)

Reply:
We have siglet validation in `receive_siglets ` - there siglets and nonces all must have consistent indices.

Reply:
Yup - basically, the precommitments, commitments, and siglets all need to be received in the same per-party order, or else the protocol will fail at some step. I don't think it matters what order it is, as long as it is consistent for all messages.